packages/azure-storage/README.md

```diff
@@ -0,0 +1,92 @@
+# language-azure-storage <img src='assets/square.png' width="30" height="30"/>
+
+An OpenFn **_adaptor_** for building integration jobs for use with the Azure
+Storage APIs. At present, this adaptor only supports the Azure Blob Storage,
+API, but could be expanded for future Azure Storage services (e.g., Files,
+Tables, Queues, etc.).
+
+## Documentation
+
+View the [docs site](https://docs.openfn.org/adaptors/packages/azure-storage-docs)
+for full technical documentation.
+
+### Configuration
+
+View the
+[configuration-schema](https://docs.openfn.org/adaptors/packages/azure-storage-configuration-schema/)
+for required and optional `configuration` properties.
+
+### An example showing multiple capabilties of this adaptor
+
+Assume you have a simple `state.json` file as follows:
+
+```json
+{
+  "data": {
+	"foo": "bar",
+	"bif": "baz",
+	"sunday": "monday",
+	"answer": "42"
+  },
+  "configuration": {
+      "accountName": "myaccountname",
+      "accountKey": "myaccountkey"
+  }
+}
+```
+
+Create a file `job.js`, as shown below, to run with the OpenFn CLI.
+
+```js
+const data = state.data;
```

Comment:
@josephjclark -- Thank you for this comment. I've reworked my example per your guidance and I think I'm very close but I'm experiencing a problem with the way that your example uses the open function to return the name of the blob. For my experiment, I'm using the following two job definitions. The first one works. The second one throws an error when trying to instantiate the `blockBlobClient` (which is the first place where the `blobName` argument is used).

This works:
```
uploadBlob(
"some/path/myblobname.json",
  state.data,
  {
    blobHTTPHeaders: { blobContentType: 'application/json' }
  },
  { createContainer: true, containerName: 'thisone'}
);
```

This doesn't work
```
uploadBlob(
  () => { return "this/myblobname.json"; },
  state.data,
  {
    blobHTTPHeaders: { blobContentType: 'application/json' }
  },
  { createContainer: true, containerName: 'thisone'}
);
```

I run the job via the CLI. When it fails, it returns the following output. The `blobName.split is not a function` error is thrown when I try to instantiate the `blockBlobClient`. I'm assuming that function hasn't yet returned, so the `blockBlobClient` is trying to instantiate itself with a function argument that causes it to error.

```
➜  nhgh_workflows openfn uploadAzureSimpler.js -ma azure-storage -s tmp/state_for_azure_upload_simple.json -O  -v
[CLI] ♦ Versions:
         ▸ node.js                           18.17.1
         ▸ cli                               0.4.10
         ▸ runtime                           0.2.1
         ▸ compiler                          0.0.38
         ▸ @openfn/language-azure-storage    latest
[CLI] ✔ Loading adaptors from monorepo at /Users/benson/sandbox/newhorizons/openfn_adaptors
[CLI] ✔ Loaded state from tmp/state_for_azure_upload_simple.json
[CLI] ✔ Compiled from uploadAzureSimple.js
[R/T] ♦ Starting job job-1
[R/T] ✘ Failed job job-1 after 154ms
[R/T] ✘ TypeError: TypeError: blobName.split is not a function
[R/T] ✘ Check state.errors.job-1 for details.
[CLI] ✔ Result: 
[CLI] ♦ {
  "data": {
    "foo": "bar",
    "bif": "baz",
    "sunday": "monday",
    "answer": "42"
  },
  "errors": {
    "job-1": {
      "type": "TypeError",
      "jobId": "job-1",
      "message": "TypeError: blobName.split is not a function",
      "error": {
        "source": "runtime",
        "type": "RuntimeError",
        "severity": "fail",
        "subtype": "TypeError",
        "name": "RuntimeError",
        "message": "TypeError: blobName.split is not a function"
      }
    }
  }
}
[CLI] ⚠ Errors reported in 1 jobs
[CLI] ✔ Finished in **231ms**
```

I'd appreciate your thoughts on this. I'd be happy to jump on a GMT call if it would help. Thank you!

Reply:
Hi @bensonmiller - that's the `expandReferences` thing biting us.

Check out this comment: https://github.com/OpenFn/adaptors/pull/438#discussion_r1410435524

You need to expand the `blobName` reference and use the "resolved" value in the function.

Reply:
Ah, yes! Thanks for the re-iteration. This should be all fixed up now.